Sources/KavaPlugin.swift

```diff
@@ -0,0 +1,256 @@
+        /// player reached 50 percent
+        case playReached50Percent = 12
+        /// player reached 75 percent
+        case playReached75Percent = 13
+        /// player reached 100 percent
+        case playReached100Percent = 14
+        /// Pause event was triggred
+        case pause = 33
+        /// Replay event was triggred
+        case replay = 34
+        /// Seeking event was triggred
+        case seek = 35
+        /// Captions event (text track was changed) was triggred
+        case captions = 38
+        /// Source Selected (media was changed) event was triggred
+        case sourceSelected = 39
+        /// Error event was triggred
+        case error = 98
+        /// Sent every 10 seconds of active playback.
+        case view = 99
+    }
+    
+    var config: KavaPluginConfig
+    var sentPlaybackPoints: [KavaEventType : Bool] = cleanPlaybackPoints()
+    var boundaryObservationToken: UUID?
+    var viewTimer: Timer?
+    var bufferingStartTime: Date?
+    var kavaData = KavaPluginData()
+    /// A sequence number which describe the order of events in a viewing session.
+    private var eventIndex = 1
+    private var sessionStartTime: String?
+    
+    /************************************************************/
+    // MARK: PKPlugin
+    /************************************************************/
+    
+    public override class var pluginName: String {
+        return "KavaPlugin"
+    }
+    
+    public required init(player: Player, pluginConfig: Any?, messageBus: MessageBus) throws {
+        guard let config = pluginConfig as? KavaPluginConfig else {
+            PKLog.error("missing plugin config or wrong plugin class type")
+            throw PKPluginError.missingPluginConfig(pluginName: KavaPlugin.pluginName)
+        }
+        
+        self.config = config
+        try super.init(player: player, pluginConfig: self.config, messageBus: messageBus)
+        self.registerEvents()
+    }
+    
+    public override func onUpdateMedia(mediaConfig: MediaConfig) {
+        PKLog.debug("onUpdateMedia: \(String(describing: mediaConfig))")
+        super.onUpdateMedia(mediaConfig: mediaConfig)
+        self.resetPlayerFlags()
+        self.unregisterFromBoundaries()
+        self.stopViewTimer()
+    }
+    
+    public override func onUpdateConfig(pluginConfig: Any) {
+        super.onUpdateConfig(pluginConfig: pluginConfig)
+        
+        guard let config = pluginConfig as? KavaPluginConfig else {
+            PKLog.error("plugin config is wrong")
+            return
+        }
+        
+        PKLog.debug("new config::\(String(describing: config))")
+        self.config = config
+    }
+    
+    public override func destroy() {
+        self.messageBus?.removeObserver(self, events: playerEventsToRegister)
+        self.unregisterFromBoundaries()
+        self.stopViewTimer()
+        super.destroy()
+    }
+    
+    static func cleanPlaybackPoints() -> [KavaEventType : Bool] {
```

Comment:
why is this static? because it's a utility belt method? I don't fully understand as we keep the data tracking as plugin instance, but the actual data as global dict?

Reply:
@vadimkononov  please answer this

Reply:
the function is static just for convenient initialiser for 'sentPlaybackPoints' property and avoiding optional on it (line 63). what is global it just the array of the points, but the dictionary with actual sent points is per instance